libsel4vm/src/arch/arm/vgic/vgic.c

```diff
@@ -151,7 +151,7 @@ struct gic_dist_map {
     uint32_t active0[CONFIG_MAX_NUM_NODES];             /* [0x300, 0x304) */
     uint32_t active[31];                                /* [0x300, 0x380) */
     uint32_t active_clr0[CONFIG_MAX_NUM_NODES];         /* [0x380, 0x384) */
-    uint32_t active_clr[32];                            /* [0x384, 0x400) */
+    uint32_t active_clr[31];                            /* [0x384, 0x400) */
```

Comment:
Seems this has always been a bug then and the comments about the offsets were never right from this place on? Are we just lucky the code every worked because nobody uses this correctly?

Reply:
It's not a bug since the only access to that field is in the distributor fault handling code, and those accesses are dependent on the offsets defined in `vgicv2_defs.h`, which are correct. I don't think it was ever possible to access the last element when it was length 32.

Reply:
I thing this was a bug, because if we have 32 words here, this uses offset 0x384 - 0x404 actually, so everything after this is 4 byte off.

Reply:
Where are we depending on these offsets?

Reply:
Maybe I'm missing something here. Is this describing some actual hardware or is this is virtual peripheral where the actual memory layout does not matter because it only ever accesses via the filed names, so the compiler handles this properly and any memory layout could be used?

Reply:
This is an emulated device so those addresses commented correspond to the actual hardware, not this register map. Since the code depends on pre-defined offsets and not the memory layout, I don't think this is a bug.

Reply:
But to answer your question, yes any memory layout could be used as long as the `#defines` in `vgicv2_defs.h` are correct.

Reply:
Let's add a 
```
compile_time_assert(gic_dist_map_size, 0x1000 == sizeo(struct gic_dist_map));
```
after the definition so the compiler can ensure things are as expected

Reply:
Since the register map is dependent on the number of VCPUs since some of the registers are banked, the size isn't always a page. But I guess I could do a slightly more complicated comptime assert.

Reply:
I wonder if we should remove all the offset comments and have a comment in the beginning that clearly states this is a pure virtual peripheral that just tries to mimic a hardware distributor. Then `CONFIG_MAX_NUM_NODES` is finde also. Otherwise the offsets there are confusing and not guaranteed. I guess this all came from a singe core implementation where this made sense, but more multiple (virtual) cores it does not.

Reply:
If this is virtual, all the "resX" could also be dropped, because nobody needs them and they don't guarantee anything when `CONFIG_MAX_NUM_NODES` changes.

Reply:
I'm implementing GICv3 right now, and was planning on adding a comment explaining why the register map is dependent on `CONFIG_MAX_NUM_NODES` and so the offsets aren't always correct. I think that they still have value, it's easier for me to go to the spec based on an offset commented rather than the name of the field.

Reply:
I'll add a comment here explaining it anyways.

Reply:
> If this is virtual, all the "resX" could also be dropped

Sure, I'll do that too.

Reply:
> offsets [...]  still have value, it's easier for me to go to the spec based on an offset commented rather than the name of the field.

Well, I'd argue here, that the specs [1][2] also use names, and these might be easier to search for than the number.

[1] https://developer.arm.com/documentation/ddi0471/b/programmers-model/distributor-register-summary
[2] https://developer.arm.com/documentation/ddi0516/e/programmers-model/distributor-register-summary